File: src/arch/x86_64/gdt.zig
const std = @import("std");
const yak = @import("../../main.zig");

const Gdtr = packed struct(u80) {
    limit: u16,
    base: u64,
};

const Dpl = enum(u2) {
    Kernel = 0,
    Ring1,
    Ring2,
    User,
};

const Access = packed struct(u8) {
    accessed: bool = false,
    rw: bool = false,
    dc: bool = false,
    executable: bool = false,
    segment: bool = false,
    dpl: Dpl = .Kernel,
    present: bool = false,
};

const GdtEntry = packed struct(u64) {
    limit: u16,
    base_low: u24,
    access: Access,
    granularity: u8,
    base_mid: u8,
};

const TssEntry = extern struct {
    common: GdtEntry align(1),
    base_high: u32 align(1),
    rsv: u32 align(1),
};

pub const GdtIndex = enum(u8) {
    Null = 0,
    KernelCode,
    KernelData,
    UserData,
    UserCode,
    Tss,

    pub fn toSeg(comptime self: @This()) comptime_int {
        return @intFromEnum(self) * 8;
    }
};

const Gdt = extern struct {
    entries: [5]GdtEntry align(1),
    tss: TssEntry align(1),

    fn makeEntry(self: *Gdt, index: GdtIndex, base: u32, limit: u16, access: Access, granularity: u8) void {
        const entry = &self.entries[@intFromEnum(index)];
        entry.limit = limit;
        entry.access = access;
        entry.granularity = granularity;
        entry.base_low = @truncate(base);
        entry.base_mid = @truncate(base >> 24);
    }

    fn makeTssEntry(self: *Gdt) void {
        self.tss.common.limit = 104;
        self.tss.common.access = .{
            .accessed = true,
            .executable = true,
            .present = true,
        };
    }
};

const Tss = packed struct {
    unused0: u32,
    rsp0: u64,
    rsp1: u64,
    rsp2: u64,
    unused1: u64,
    ist1: u64, // NMI
    ist2: u64, // debug/breakpoint
    ist3: u64, // double fault
    ist4: u64,
    ist5: u64,
    ist6: u64,
    ist7: u64,
    unused2: u64,
    iopb: u32,
};

var gdt: Gdt align(16) = undefined;

pub fn initGdt() void {
    @memset(std.mem.asBytes(&gdt), 0);
    gdt.makeEntry(.Null, 0, 0, .{}, 0);
    gdt.makeEntry(.KernelCode, 0, 0, .{
        .accessed = true,
        .rw = true,
        .dc = false,
        .executable = true,
        .segment = true,
        .dpl = .Kernel,
        .present = true,
    }, 0x20);
    gdt.makeEntry(.KernelData, 0, 0, .{
        .accessed = true,
        .rw = true,
        .dc = false,
        .executable = false,
        .segment = true,
        .dpl = .Kernel,
        .present = true,
    }, 0);
    gdt.makeEntry(.UserCode, 0, 0, .{
        .accessed = true,
        .rw = true,
        .dc = false,
        .executable = true,
        .segment = true,
        .dpl = .User,
        .present = true,
    }, 0x20);
    gdt.makeEntry(.UserData, 0, 0, .{
        .accessed = true,
        .rw = true,
        .dc = false,
        .executable = false,
        .segment = true,
        .dpl = .User,
        .present = true,
    }, 0);
    gdt.makeTssEntry();
}

pub fn loadGdt() void {
    var gdtr: Gdtr = .{
        .base = @intFromPtr(&gdt),
        .limit = @sizeOf(Gdt) - 1,
    };

    asm volatile ("lgdt %[p]"
        :
        : [p] "*p" (&gdtr.limit),
    );

    asm volatile (
        \\pushq %[cs]
        \\leaq 1f(%%rip), %%rax
        \\pushq %%rax
        \\lretq
        \\1:
        \\mov %[ds], %%ds
        \\mov %[ds], %%es
        \\mov %[ds], %%ss
        \\mov %[ds], %%gs
        \\mov %[ds], %%fs
        :
        : [cs] "i" (GdtIndex.KernelCode.toSeg()),
          [ds] "rm" (GdtIndex.KernelData.toSeg()),
        : "rax"
    );
}

var tsslock: yak.sync.Spinlock = .init();

pub fn loadTss(tss: *Tss) void {
    const ints = tsslock.lockInts();
    defer tsslock.unlockInts(ints);

    const tss_addr: u64 = @bitCast(tss);
    gdt.tss.common.base_low = @truncate(tss_addr);
    gdt.tss.common.base_mid = @truncate(tss_addr >> 24);
    gdt.tss.common.flags = 0;
    gdt.tss.common.access = .{
        .present = true,
        .executable = true,
        .accessed = true,
    };

    gdt.tss.base_high = @truncate(tss_addr >> 32);
    gdt.tss.rsv = 0;

    asm volatile ("ltr %[p]"
        :
        : [p] "rm16" (GdtIndex.Tss.toSeg()),
        : "memory"
    );
}

File: kernel/arch/x86_64/src/gdt.c
#include <stdint.h>
#include <yak/vm.h>
#include <yak/vm/map.h>
#include <yak/percpu.h>
#include <yak/cpudata.h>
#include <string.h>

#include "gdt.h"
#include "tss.h"

struct [[gnu::packed]] gdt_entry {
	uint16_t limit;
	uint16_t low;
	uint8_t mid;
	uint8_t access;
	uint8_t granularity;
	uint8_t high;
};

struct [[gnu::packed]] gdt_tss_entry {
	uint16_t length;
	uint16_t low;
	uint8_t mid;
	uint8_t access;
	uint8_t flags;
	uint8_t high;
	uint32_t upper;
	uint32_t rsv;
};

struct [[gnu::packed]] gdt {
	struct gdt_entry entries[GDT_TSS];
	struct gdt_tss_entry tss;
};

__percpu struct gdt t_gdt;
__percpu struct tss t_tss;

static void gdt_make_entry(int index, uint32_t base, uint16_t limit,
			   uint8_t access, uint8_t granularity)
{
	t_gdt.entries[index].limit = limit;
	t_gdt.entries[index].access = access;
	t_gdt.entries[index].granularity = granularity;
	t_gdt.entries[index].low = (uint16_t)base;
	t_gdt.entries[index].mid = (uint8_t)(base >> 16);
	t_gdt.entries[index].high = (uint8_t)(base >> 24);
}

static void gdt_tss_entry()
{
	t_gdt.tss.length = 104;
	t_gdt.tss.access = 0x89;
	t_gdt.tss.flags = 0;
	t_gdt.tss.rsv = 0;

	t_gdt.tss.low = 0;
	t_gdt.tss.mid = 0;
	t_gdt.tss.high = 0;
	t_gdt.tss.upper = 0;
}

void gdt_init()
{
	gdt_make_entry(GDT_NULL, 0, 0, 0, 0);
	gdt_make_entry(GDT_KERNEL_CODE, 0, 0, 0x9a, 0x20);
	gdt_make_entry(GDT_KERNEL_DATA, 0, 0, 0x92, 0);
	gdt_make_entry(GDT_USER_DATA, 0, 0, 0xf2, 0);
	gdt_make_entry(GDT_USER_CODE, 0, 0, 0xfa, 0x20);
	gdt_tss_entry();
}

static vaddr_t alloc_kstack()
{
	vaddr_t stack_addr;
	EXPECT(vm_map(kmap(), NULL, KSTACK_SIZE, 0, 0, VM_RW | VM_PREFILL,
		      VM_INHERIT_NONE, &stack_addr));
	return stack_addr + KSTACK_SIZE;
}

static void tss_reload()
{
	uint64_t tss_addr = (uint64_t)PERCPU_PTR(void, t_tss);
	t_gdt.tss.low = (uint16_t)tss_addr;
	t_gdt.tss.mid = (uint8_t)(tss_addr >> 16);
	t_gdt.tss.high = (uint8_t)(tss_addr >> 24);
	t_gdt.tss.upper = (uint32_t)(tss_addr >> 32);

	t_gdt.tss.flags = 0;
	t_gdt.tss.access = 0b10001001;
	t_gdt.tss.rsv = 0;
	asm volatile("ltr %0" ::"rm"(GDT_SEL_TSS) : "memory");
}

void tss_set_rsp0(uint64_t stack_top)
{
	t_tss.rsp0 = stack_top;
}

void tss_init()
{
	memset(PERCPU_PTR(void, t_tss), 0, sizeof(struct tss));
	t_tss.ist1 = alloc_kstack();
	t_tss.ist2 = alloc_kstack();
	t_tss.ist3 = alloc_kstack();
	t_tss.ist4 = alloc_kstack();
	tss_reload();
}

void gdt_reload()
{
	struct [[gnu::packed]] {
		uint16_t limit;
		uint64_t base;
	} gdtr = {
		.limit = sizeof(struct gdt) - 1,
		.base = (uint64_t)PERCPU_PTR(struct gdt, t_gdt),
	};

	asm volatile( //
		"lgdt (%0)\n\t"
		// reload CS
		"pushq %1\n\t"
		"leaq 1f(%%rip), %%rax\n\t"
		"pushq %%rax\n\t"
		"lretq\n\t"
		"1:\n\t"
		// reload DS
		"mov %2, %%ds\n\t"
		"mov %2, %%es\n\t"
		"mov %2, %%ss\n\t"
		"xor %%eax, %%eax\n\t"
		"mov %%ax, %%fs\n\t"
		//
		:
		: "r"(&gdtr), "i"((uint16_t)GDT_SEL_KERNEL_CODE),
		  "r"(GDT_SEL_KERNEL_DATA)
		: "rax", "memory"
		//
	);
}
